Test failure: DCMotor dynamics not working

## Test Failure Description

**Test:** `TestDCMotor.test_motor_dynamics`
**File:** `robocar/simulation/tests/test_robot_model.py:60`

### Error Details
```
assert omega > 0
AssertionError: assert 0.0 > 0

tests/test_robot_model.py:60: AssertionError
```

### Expected Behavior
When voltage is applied to the DC motor model, it should start spinning and `omega` (angular velocity) should be greater than 0.

### Actual Behavior
The motor angular velocity remains 0.0 even after applying voltage and updating with load torque.

### Test Configuration
- Voltage: 5.0V
- Load torque: 0.1
- Time step: 0.01s
- Motor config includes resistance, inductance, back EMF constant, etc.

### Root Cause
The DCMotor.update() method may not be correctly implementing the motor dynamics equations or the motor state is not being properly updated over time.

### Location
`robocar/simulation/src/robot_model.py` - DCMotor class

--- a/ESP-IDF_Robot_Simulation.md
+++ b/ESP-IDF_Robot_Simulation.md
@@ -1,0 +1,75 @@
+# ESP-IDF Robot Simulation: Comprehensive Development Guide
+
+The landscape of 3D physics-based simulation for ESP-IDF robot car firmware has evolved dramatically, offering multiple viable approaches from browser-based prototyping to professional-grade physics engines. **Current solutions achieve 1kHz control loops with \<5% accuracy** compared to real hardware, transforming embedded robotics development from hardware-dependent to simulation-first workflows.
+
+Modern ESP-IDF simulation encompasses three primary approaches: **direct ESP32 emulation** through QEMU and FreeRTOS POSIX ports, **hardware-in-the-loop integration** connecting real ESP32s to virtual environments, and **abstraction layer strategies** enabling cross-compilation to host systems. Professional implementations demonstrate successful deployment in commercial quadruped robots and educational institutions worldwide.
+
+## Open-source robotics platforms lead integration capabilities
+
+**Gazebo dominates professional robotics simulation** with its mature ecosystem and robust ROS2 integration. The transition from Gazebo Classic to the unified Gazebo (formerly Ignition) platform provides enhanced physics accuracy through ODE, Bullet, and DART engines. ESP-IDF integration operates through **plugin-based communication bridges** supporting MQTT, WebSocket, and direct TCP/UDP protocols.
+
+Technical implementation centers on custom Gazebo plugins interfacing with ESP32 hardware. A typical integration creates ModelPlugin or SensorPlugin components that establish pub/sub messaging via `gz-transport` using Protocol Buffers serialization. **Communication latency achieves 0.2ms via Ethernet** and 1.2ms through WiFi ESP-NOW, enabling real-time control applications. The plugin architecture supports bidirectional sensor data streaming and actuator command processing.
+
+**Webots offers superior educational integration** with its open-source Apache 2.0 license and comprehensive documentation. The platform excels at **external controller integration** where ESP32 devices connect via TCP/UDP sockets while Webots handles physics simulation. The `webots_ros2_driver` package enables micro-ROS communication, allowing ESP32s to participate in ROS2 ecosystems. Webots provides deterministic simulation with configurable time steps, crucial for reproducible embedded system testing.
+
+**PyBullet presents the most flexible Python-based approach** with excellent performance characteristics achieving 10-100x real-time simulation speeds. Integration strategies focus on **RESTful API communication** where ESP32 devices send HTTP requests to PyBullet servers, or MQTT publish/subscribe patterns for sensor data exchange. The framework supports GPU acceleration through CUDA and provides efficient collision detection suitable for complex robot car environments.
+
+## ESP-IDF simulation achieves production-ready capabilities
+
+**Official QEMU integration** represents Espressif's commitment to simulation-based development. The maintained QEMU fork supports ESP32 CPU emulation with virtual framebuffer devices and GDB debugging integration. While still evolving, **`idf.py qemu` commands** enable direct execution of ESP-IDF applications in emulated environments. Current limitations include incomplete peripheral emulation and timing accuracy constraints.
+
+**Community simulators provide immediate practical value**. Wokwi's browser-based platform supports real-time ESP32 simulation with WiFi capabilities and custom firmware upload. The platform handles ESP32, ESP32-S2, ESP32-S3, ESP32-C3, ESP32-C6, and ESP32-H2 variants with integrated sensor libraries. **VS Code extension integration** enables seamless development workflows.
+
+**Hardware-in-the-loop approaches** bridge simulation and reality by connecting physical ESP32s to virtual environments. This strategy preserves real-time hardware behavior while enabling complex environment simulation. Professional HIL systems from dSPACE and National Instruments integrate ESP32 boards with real-time simulation, while **DIY approaches** use Raspberry Pi hosts with serial communication for cost-effective testing.
+
+The **FreeRTOS POSIX simulator** enables ESP-IDF applications to run on Linux hosts through `idf.py --preview set-target linux`. This approach maintains FreeRTOS task scheduling behavior while providing host-based debugging capabilities. Signal-based tick interrupts approximate real-time constraints, though **timing accuracy remains limited** for safety-critical applications.
+
+## Physics simulation accuracy depends on engine selection and implementation
+
+**Motor control simulation** requires mathematical models combining electrical and mechanical dynamics. DC motor behavior follows state-space representations incorporating back-EMF, inductance, resistance, and mechanical inertia. **Advanced models include nonlinear effects** like saturation, cogging torque, and temperature dependencies. Simulation accuracy typically achieves ±0.1° position accuracy for servo applications with proper parameter identification.
+
+**Differential drive kinematics** form the foundation of robot car simulation. Forward and inverse kinematic equations convert wheel velocities to robot motion, while **dynamic models incorporate actuator dynamics** and environmental forces. Recent research indicates velocity-based models outperform traditional torque-based approaches for mobile robot simulation.
+
+**Encoder feedback simulation** generates quadrature signals with configurable noise characteristics. Implementation includes electrical noise (±0.1-1 LSB), mechanical vibration components, and temperature drift effects. **Timing accuracy maintains microsecond precision** for realistic encoder behavior, crucial for closed-loop control validation.
+
+**Wheel-ground interaction physics** determines robot car behavior through friction and slip modeling. The Pacejka Magic Formula provides industry-standard tire dynamics, while simpler Coulomb friction models suffice for basic applications. **Physics engine selection significantly impacts accuracy**: Bullet excels at slip simulation, MuJoCo provides superior contact dynamics, while PhysX offers hardware acceleration for large-scale simulations.
+
+## Advanced sensor simulation enables comprehensive testing
+
+**IMU sensor modeling** incorporates Allan variance analysis to simulate realistic noise characteristics. Accelerometer and gyroscope signals include angle random walk, bias instability, and rate random walk components. **MEMS sensor characteristics** match commercial specifications with typical accelerometer noise at 100μg/√Hz and gyroscope noise at 0.01°/s/√Hz.
+
+**Ultrasonic sensor physics** simulates wave propagation with realistic beam patterns and multipath effects. Time-of-flight calculations account for temperature-dependent sound speed variations and environmental noise. **Implementation includes** quantization effects, thermal drift, and interference between multiple sensors.
+
+**Camera vision simulation** provides synthetic datasets for computer vision algorithm development. Virtual camera implementations support configurable intrinsic/extrinsic parameters, lens distortion models, and lighting conditions. **OpenCV integration** enables real-time image processing pipeline testing with controlled environmental parameters.
+
+**Communication protocol simulation** models I2C, SPI, and UART timing characteristics with microsecond accuracy. Implementation includes bus arbitration, error detection mechanisms, and multi-MCU communication scenarios. **Timing-accurate simulation** maintains protocol specifications for reliable embedded system validation.
+
+## Custom development frameworks offer maximum flexibility
+
+**Unity Robotics Hub** provides comprehensive simulation capabilities with URDF import support and two-way ROS2 communication. The platform enables **synthetic dataset generation** for machine learning training while maintaining real-time performance exceeding 60 FPS. ArticulationBody physics system offers accurate joint dynamics simulation suitable for complex robotic systems.
+
+**Unreal Engine integration** through rclUE plugins enables photorealistic simulation with Chaos Physics engine. The Unreal Robotics Lab framework incorporates MuJoCo physics for research-grade accuracy. **Lumen global illumination** provides realistic lighting conditions for computer vision testing, while Blueprint visual scripting accelerates prototyping.
+
+**Python-based custom frameworks** offer maximum ESP-IDF integration flexibility. The Robotics Toolbox for Python with Swift visualizer supports comprehensive robot modeling and path planning. **NumPy/SciPy integration** enables efficient custom physics simulation with real-time performance constraints achievable through optimized algorithms.
+
+**Web-based solutions** using WebGL/Three.js provide cross-platform accessibility without installation requirements. Browser-based simulation enables **direct WebSocket communication** with ESP32 devices, supporting real-time sensor data streaming and command processing. JavaScript physics engines like Cannon.js and Ammo.js provide adequate simulation accuracy for educational and prototyping applications.
+
+## Real-world validation demonstrates practical effectiveness
+
+**Academic institutions** successfully integrate ESP-IDF simulation into engineering curricula. The University of Ljubljana's IoT educational framework demonstrates positive learning outcomes using ESP32-based robotics projects. **Student projects** include weather stations, home automation systems, and mobile robots with comprehensive simulation validation.
+
+**Commercial applications** achieve production-ready implementations. The Open Dynamic Robot Initiative's master board project demonstrates **1kHz real-time control** with 0.2ms Ethernet latency in professional quadruped robots. Educational platforms like the Society of Robotics and Automation's SRA Board provide comprehensive development environments for undergraduate robotics courses.
+
+**Simulation accuracy studies** show promising results. Validation research comparing CoppeliaSim, Gazebo, MORSE, and Webots against real Husky A200 robot data indicates **\<5% error** in force/position tracking for properly configured systems. Soft robotics validation using Webots simulation of pneumatic grippers demonstrates high-fidelity modeling capabilities.
+
+## Implementation recommendations prioritize practical deployment
+
+**Platform selection depends on application requirements**. Educational and research applications benefit from Webots' comprehensive documentation and ROS2 integration. High-fidelity physics demands favor Gazebo with custom plugins, while rapid prototyping succeeds with PyBullet's Python flexibility. **Production applications** require careful evaluation of accuracy versus performance trade-offs.
+
+**Integration architecture should emphasize modularity**. Recommended communication flows connect ESP-IDF robots through WebSocket protocols to simulation engines with physics engine backends. **JSON message formats** provide structured data exchange while binary protocols handle high-frequency sensor data efficiently.
+
+**Development workflow integration** enables continuous validation. Automated CI/CD pipelines combine simulation-based testing with hardware validation, ensuring reliability across development cycles. **pytest-embedded frameworks** support multi-target testing with automatic binary generation and artifact management.
+
+**Migration strategies** should begin with software-only components using incremental approaches. Hardware abstraction through CMock enables gradual transition from physical to simulated testing. **Validation requirements** mandate maintaining physical testing for safety-critical components while accelerating development through simulation.
+
+The convergence of mature simulation platforms, robust ESP-IDF integration capabilities, and validated real-world deployments creates unprecedented opportunities for embedded robotics development. Success requires careful platform selection, thorough validation methodologies, and acknowledgment of simulation limitations while leveraging the significant advantages of simulation-first development workflows.

--- a/robocar/.gitignore
+++ b/robocar/.gitignore
@@ -1,3 +1,9 @@
 build
 cam/credentials.h
+esp32-cam-idf/main/credentials.h
 .aider*
+# Python build artifacts
+simulation/src/__pycache__/
+*.pyc
+*.pyo
+*.pyd

--- a/robocar/.mcp.json
+++ b/robocar/.mcp.json
@@ -1,0 +1,11 @@
+{
+  "mcpServers": {
+    "zen": {
+      "command": "sh",
+      "args": [
+        "-c",
+        "exec $(which uvx || echo uvx) --from git+https://github.com/BeehiveInnovations/zen-mcp-server.git zen-mcp-server"
+      ]
+    }
+  }
+}

--- a/robocar/esp32-cam-idf/main/CMakeLists.txt
+++ b/robocar/esp32-cam-idf/main/CMakeLists.txt
@@ -1,3 +1,3 @@
-idf_component_register(SRCS "main.c" "camera.c" "wifi_manager.c" "claude_api.c" "base64.c" "ai_response_parser.c" "i2c_master.c" "i2c_protocol.c" "ai_backend.c" "ollama_backend.c" "ollama_discovery.c"
+idf_component_register(SRCS "main.c" "camera.c" "wifi_manager.c" "claude_api.c" "base64.c" "ai_response_parser.c" "i2c_master.c" "i2c_protocol.c" "ai_backend.c" "ollama_backend.c" "ollama_discovery.c" "credentials_loader.c"
                     INCLUDE_DIRS "."
                     REQUIRES "esp32-camera" "esp_wifi" "esp_http_client" "json" "nvs_flash" "driver" "espressif__mdns")

--- a/robocar/esp32-cam-idf/main/ai_response_parser.c
+++ b/robocar/esp32-cam-idf/main/ai_response_parser.c
@@ -106,8 +106,42 @@ bool parse_ai_response(const char* response_text, ai_command_t* command) {
     cJSON_Delete(root);
 
     if (!json_obj) {
-        ESP_LOGE(TAG, "Failed to parse extracted JSON");
-        return false;
+        ESP_LOGE(TAG, "Failed to parse extracted JSON: %s", cJSON_GetErrorPtr() ? cJSON_GetErrorPtr() : "Unknown error");
+        
+        // Try to extract simple keywords from text as fallback
+        ESP_LOGW(TAG, "Attempting fallback keyword extraction");
+        if (strstr(content_text, "forward") || strstr(content_text, "move forward")) {
+            strcpy(command->movement_command, "F");
+            command->has_movement = true;
+            ESP_LOGI(TAG, "Fallback: detected 'forward' command");
+            return true;
+        } else if (strstr(content_text, "backward") || strstr(content_text, "move backward")) {
+            strcpy(command->movement_command, "B");
+            command->has_movement = true;
+            ESP_LOGI(TAG, "Fallback: detected 'backward' command");
+            return true;
+        } else if (strstr(content_text, "left") || strstr(content_text, "turn left")) {
+            strcpy(command->movement_command, "L");
+            command->has_movement = true;
+            ESP_LOGI(TAG, "Fallback: detected 'left' command");
+            return true;
+        } else if (strstr(content_text, "right") || strstr(content_text, "turn right")) {
+            strcpy(command->movement_command, "R");
+            command->has_movement = true;
+            ESP_LOGI(TAG, "Fallback: detected 'right' command");
+            return true;
+        } else if (strstr(content_text, "stop") || strstr(content_text, "halt")) {
+            strcpy(command->movement_command, "S");
+            command->has_movement = true;
+            ESP_LOGI(TAG, "Fallback: detected 'stop' command");
+            return true;
+        }
+        
+        // Default fallback: stop command for safety
+        strcpy(command->movement_command, "S");
+        command->has_movement = true;
+        ESP_LOGW(TAG, "Fallback: using default 'stop' command for safety");
+        return true;
     }
 
     bool success = false;
@@ -193,6 +227,15 @@ bool parse_ai_response(const char* response_text, ai_command_t* command) {
         command->display_line = 6; // Default to line 6 for Claude messages
     }
 
+    // Cleanup JSON object
     cJSON_Delete(json_obj);
+    
+    // Log parsing results
+    if (success) {
+        ESP_LOGI(TAG, "Successfully parsed AI response");
+    } else {
+        ESP_LOGW(TAG, "AI response parsed but no valid commands found");
+    }
+    
     return success;
 }

--- a/robocar/esp32-cam-idf/main/claude_api.c
+++ b/robocar/esp32-cam-idf/main/claude_api.c
@@ -33,10 +33,23 @@ static esp_err_t http_event_handler(esp_http_client_event_t *evt) {
             break;
         case HTTP_EVENT_ON_DATA:
             ESP_LOGD(TAG, "HTTP_EVENT_ON_DATA, len=%d", evt->data_len);
-            if (s_response_buffer && s_response_len + evt->data_len < 4096 - 1) {
-                memcpy(s_response_buffer + s_response_len, evt->data, evt->data_len);
-                s_response_len += evt->data_len;
-                s_response_buffer[s_response_len] = '\0';
+            if (s_response_buffer && evt->data && evt->data_len > 0) {
+                // Ensure we don't overflow the buffer (4096 bytes total, -1 for null terminator)
+                size_t available_space = 4095 - s_response_len;
+                size_t bytes_to_copy = (evt->data_len <= available_space) ? evt->data_len : available_space;
+                
+                if (bytes_to_copy > 0) {
+                    memcpy(s_response_buffer + s_response_len, evt->data, bytes_to_copy);
+                    s_response_len += bytes_to_copy;
+                    s_response_buffer[s_response_len] = '\0';
+                    
+                    if (bytes_to_copy < evt->data_len) {
+                        ESP_LOGW(TAG, "HTTP response truncated: received %d bytes, only %zu bytes available", 
+                                evt->data_len, available_space);
+                    }
+                } else {
+                    ESP_LOGW(TAG, "HTTP response buffer full, discarding %d bytes", evt->data_len);
+                }
             }
             break;
         case HTTP_EVENT_ON_FINISH:

--- a/robocar/esp32-cam-idf/main/config.h
+++ b/robocar/esp32-cam-idf/main/config.h
@@ -6,8 +6,9 @@
 #ifndef CONFIG_H
 #define CONFIG_H
 
-// WiFi Configuration - Include credentials from credentials.h
-#include "credentials.h"
+// WiFi Configuration - Use secure credential loader
+// Note: Credentials are loaded dynamically via credentials_loader module
+// This supports both environment variables and credentials.h fallback
 
 // Timing Configuration
 #define CAPTURE_INTERVAL_MS     5000    // Capture image every 5 seconds

--- a/robocar/esp32-cam-idf/main/credentials.h.template
+++ b/robocar/esp32-cam-idf/main/credentials.h.template
@@ -1,0 +1,36 @@
+/**
+ * @file credentials.h.template
+ * @brief Template for WiFi and API credentials - SECURITY CRITICAL
+ *
+ * SECURITY INSTRUCTIONS:
+ * 1. Copy this file to credentials.h 
+ * 2. Fill in your actual credentials
+ * 3. NEVER commit credentials.h to the repository
+ * 4. Use environment variables for CI/CD deployments
+ * 
+ * IMPORTANT: credentials.h is in .gitignore to prevent accidental commits
+ */
+
+#ifndef CREDENTIALS_H
+#define CREDENTIALS_H
+
+// WiFi Configuration
+// Replace with your actual WiFi network credentials
+#define WIFI_SSID "YOUR_WIFI_SSID_HERE"
+#define WIFI_PASSWORD "YOUR_WIFI_PASSWORD_HERE"
+
+// Claude API Configuration  
+// Get your API key from: https://console.anthropic.com/
+// CRITICAL: Rotate any exposed keys immediately
+#define CLAUDE_API_KEY "sk-ant-api03-YOUR_ACTUAL_API_KEY_HERE"
+
+// Security Notes:
+// - Never share these credentials in public repositories
+// - Rotate API keys regularly (recommended: every 90 days)
+// - Use different credentials for development vs production
+// - For production deployment, use environment variables:
+//   export CLAUDE_API_KEY="your-key"
+//   export WIFI_SSID="your-ssid" 
+//   export WIFI_PASSWORD="your-password"
+
+#endif // CREDENTIALS_H

--- a/robocar/esp32-cam-idf/main/credentials_loader.c
+++ b/robocar/esp32-cam-idf/main/credentials_loader.c
@@ -1,0 +1,202 @@
+/**
+ * @file credentials_loader.c
+ * @brief Secure credential loading implementation
+ */
+
+#include "credentials_loader.h"
+#include "esp_log.h"
+#include <string.h>
+#include <stdlib.h>
+
+static const char *TAG = "credentials_loader";
+
+// Global credentials instance
+static credentials_t g_credentials = {0};
+static bool g_credentials_initialized = false;
+
+/**
+ * @brief Attempt to load credential from environment variable
+ * 
+ * @param env_var Environment variable name
+ * @param dest Destination buffer
+ * @param max_len Maximum length of destination buffer
+ * @return true if loaded successfully, false otherwise
+ */
+static bool load_from_env(const char* env_var, char* dest, size_t max_len) {
+    const char* value = getenv(env_var);
+    if (value && strlen(value) > 0 && strlen(value) < max_len) {
+        strncpy(dest, value, max_len - 1);
+        dest[max_len - 1] = '\0';
+        ESP_LOGI(TAG, "Loaded %s from environment variable", env_var);
+        return true;
+    }
+    return false;
+}
+
+/**
+ * @brief Load credentials from credentials.h file
+ * 
+ * @param creds Pointer to credentials structure
+ * @return true if loaded successfully, false otherwise
+ */
+static bool load_from_file(credentials_t* creds) {
+    // Check if credentials.h exists and contains valid data
+    #ifdef WIFI_SSID
+    if (strlen(WIFI_SSID) > 0 && strcmp(WIFI_SSID, "YOUR_WIFI_SSID_HERE") != 0) {
+        strncpy(creds->wifi_ssid, WIFI_SSID, MAX_SSID_LENGTH - 1);
+        creds->wifi_ssid[MAX_SSID_LENGTH - 1] = '\0';
+        ESP_LOGI(TAG, "Loaded WiFi SSID from credentials.h");
+    } else {
+        ESP_LOGW(TAG, "WiFi SSID not configured in credentials.h");
+        return false;
+    }
+    #else
+    ESP_LOGW(TAG, "WIFI_SSID not defined in credentials.h");
+    return false;
+    #endif
+
+    #ifdef WIFI_PASSWORD
+    if (strlen(WIFI_PASSWORD) > 0 && strcmp(WIFI_PASSWORD, "YOUR_WIFI_PASSWORD_HERE") != 0) {
+        strncpy(creds->wifi_password, WIFI_PASSWORD, MAX_PASSWORD_LENGTH - 1);
+        creds->wifi_password[MAX_PASSWORD_LENGTH - 1] = '\0';
+        ESP_LOGI(TAG, "Loaded WiFi password from credentials.h");
+    } else {
+        ESP_LOGW(TAG, "WiFi password not configured in credentials.h");
+        return false;
+    }
+    #else
+    ESP_LOGW(TAG, "WIFI_PASSWORD not defined in credentials.h");
+    return false;
+    #endif
+
+    #ifdef CLAUDE_API_KEY
+    if (strlen(CLAUDE_API_KEY) > 0 && strcmp(CLAUDE_API_KEY, "sk-ant-api03-YOUR_ACTUAL_API_KEY_HERE") != 0) {
+        strncpy(creds->claude_api_key, CLAUDE_API_KEY, MAX_API_KEY_LENGTH - 1);
+        creds->claude_api_key[MAX_API_KEY_LENGTH - 1] = '\0';
+        ESP_LOGI(TAG, "Loaded Claude API key from credentials.h");
+    } else {
+        ESP_LOGW(TAG, "Claude API key not configured in credentials.h");
+        // API key is optional for Ollama backend
+    }
+    #else
+    ESP_LOGW(TAG, "CLAUDE_API_KEY not defined in credentials.h");
+    // API key is optional for Ollama backend
+    #endif
+
+    return true;
+}
+
+bool load_credentials(credentials_t* creds) {
+    if (!creds) {
+        ESP_LOGE(TAG, "Invalid credentials pointer");
+        return false;
+    }
+
+    // Clear credentials structure
+    memset(creds, 0, sizeof(credentials_t));
+
+    ESP_LOGI(TAG, "Loading credentials...");
+
+    // Try environment variables first
+    bool env_loaded = true;
+    if (!load_from_env("WIFI_SSID", creds->wifi_ssid, MAX_SSID_LENGTH)) {
+        env_loaded = false;
+    }
+    if (!load_from_env("WIFI_PASSWORD", creds->wifi_password, MAX_PASSWORD_LENGTH)) {
+        env_loaded = false;
+    }
+    // Claude API key is optional (for Ollama backend)
+    load_from_env("CLAUDE_API_KEY", creds->claude_api_key, MAX_API_KEY_LENGTH);
+
+    if (env_loaded) {
+        ESP_LOGI(TAG, "Successfully loaded credentials from environment variables");
+        creds->credentials_loaded = true;
+        return true;
+    }
+
+    ESP_LOGI(TAG, "Environment variables not available, trying credentials.h file...");
+
+    // Fallback to credentials.h file
+    #include "credentials.h"
+    
+    if (load_from_file(creds)) {
+        ESP_LOGI(TAG, "Successfully loaded credentials from credentials.h");
+        creds->credentials_loaded = true;
+        return true;
+    }
+
+    ESP_LOGE(TAG, "Failed to load credentials from any source");
+    return false;
+}
+
+bool validate_credentials(const credentials_t* creds) {
+    if (!creds || !creds->credentials_loaded) {
+        ESP_LOGE(TAG, "No credentials loaded");
+        return false;
+    }
+
+    // Validate WiFi credentials (required)
+    if (strlen(creds->wifi_ssid) == 0) {
+        ESP_LOGE(TAG, "WiFi SSID is empty");
+        return false;
+    }
+
+    if (strlen(creds->wifi_password) == 0) {
+        ESP_LOGE(TAG, "WiFi password is empty");
+        return false;
+    }
+
+    // Claude API key validation (optional - only required for Claude backend)
+    #ifdef CONFIG_AI_BACKEND_CLAUDE
+    if (strlen(creds->claude_api_key) == 0) {
+        ESP_LOGE(TAG, "Claude API key is required for Claude backend but not provided");
+        return false;
+    }
+
+    // Basic API key format validation
+    if (strncmp(creds->claude_api_key, "sk-ant-api03-", 13) != 0) {
+        ESP_LOGE(TAG, "Claude API key format appears invalid");
+        return false;
+    }
+    #endif
+
+    ESP_LOGI(TAG, "Credentials validation successful");
+    return true;
+}
+
+bool are_credentials_available(void) {
+    if (!g_credentials_initialized) {
+        g_credentials_initialized = true;
+        if (!load_credentials(&g_credentials)) {
+            return false;
+        }
+        if (!validate_credentials(&g_credentials)) {
+            return false;
+        }
+    }
+    return g_credentials.credentials_loaded;
+}
+
+const char* get_wifi_ssid(void) {
+    if (!are_credentials_available()) {
+        ESP_LOGE(TAG, "Credentials not available");
+        return NULL;
+    }
+    return g_credentials.wifi_ssid;
+}
+
+const char* get_wifi_password(void) {
+    if (!are_credentials_available()) {
+        ESP_LOGE(TAG, "Credentials not available");
+        return NULL;
+    }
+    return g_credentials.wifi_password;
+}
+
+const char* get_claude_api_key(void) {
+    if (!are_credentials_available()) {
+        ESP_LOGE(TAG, "Credentials not available");
+        return NULL;
+    }
+    return g_credentials.claude_api_key;
+}

--- a/robocar/esp32-cam-idf/main/credentials_loader.h
+++ b/robocar/esp32-cam-idf/main/credentials_loader.h
@@ -1,0 +1,79 @@
+/**
+ * @file credentials_loader.h
+ * @brief Secure credential loading with environment variable support
+ * 
+ * This module provides secure credential loading that:
+ * 1. Checks environment variables first (for CI/CD)
+ * 2. Falls back to credentials.h file
+ * 3. Provides validation and secure defaults
+ */
+
+#ifndef CREDENTIALS_LOADER_H
+#define CREDENTIALS_LOADER_H
+
+#include <stdbool.h>
+
+// Maximum lengths for credentials
+#define MAX_SSID_LENGTH 32
+#define MAX_PASSWORD_LENGTH 64
+#define MAX_API_KEY_LENGTH 256
+
+/**
+ * @brief Structure to hold loaded credentials
+ */
+typedef struct {
+    char wifi_ssid[MAX_SSID_LENGTH];
+    char wifi_password[MAX_PASSWORD_LENGTH];
+    char claude_api_key[MAX_API_KEY_LENGTH];
+    bool credentials_loaded;
+} credentials_t;
+
+/**
+ * @brief Load credentials from environment variables or credentials.h
+ * 
+ * This function attempts to load credentials in the following order:
+ * 1. Environment variables (WIFI_SSID, WIFI_PASSWORD, CLAUDE_API_KEY)
+ * 2. Fallback to credentials.h file
+ * 
+ * @param creds Pointer to credentials structure to populate
+ * @return true if credentials were loaded successfully, false otherwise
+ */
+bool load_credentials(credentials_t* creds);
+
+/**
+ * @brief Validate loaded credentials
+ * 
+ * @param creds Pointer to credentials structure to validate
+ * @return true if credentials are valid, false otherwise
+ */
+bool validate_credentials(const credentials_t* creds);
+
+/**
+ * @brief Get WiFi SSID
+ * 
+ * @return Pointer to WiFi SSID string
+ */
+const char* get_wifi_ssid(void);
+
+/**
+ * @brief Get WiFi password
+ * 
+ * @return Pointer to WiFi password string
+ */
+const char* get_wifi_password(void);
+
+/**
+ * @brief Get Claude API key
+ * 
+ * @return Pointer to Claude API key string
+ */
+const char* get_claude_api_key(void);
+
+/**
+ * @brief Check if credentials are available
+ * 
+ * @return true if credentials are loaded and valid, false otherwise
+ */
+bool are_credentials_available(void);
+
+#endif // CREDENTIALS_LOADER_H

--- a/robocar/esp32-cam-idf/main/main.c
+++ b/robocar/esp32-cam-idf/main/main.c
@@ -20,6 +20,7 @@
 #include "i2c_master.h"
 #include "ai_response_parser.h"
 #include "camera_pins.h"
+#include "credentials_loader.h"
 
 static const char *TAG = "esp32-cam-robocar";
 
@@ -77,9 +78,20 @@ void app_main(void) {
         return;
     }
 
-    // Connect to WiFi
-    ESP_LOGI(TAG, "Connecting to WiFi: %s", WIFI_SSID);
-    if (wifi_connect(WIFI_SSID, WIFI_PASSWORD) == ESP_OK) {
+    // Load and validate credentials
+    ESP_LOGI(TAG, "Loading credentials...");
+    if (!are_credentials_available()) {
+        ESP_LOGE(TAG, "Failed to load or validate credentials");
+        i2c_send_sound_command(SOUND_ALERT); // Send alert sound
+        return;
+    }
+
+    // Connect to WiFi using loaded credentials
+    const char* wifi_ssid = get_wifi_ssid();
+    const char* wifi_password = get_wifi_password();
+    
+    ESP_LOGI(TAG, "Connecting to WiFi: %s", wifi_ssid);
+    if (wifi_connect(wifi_ssid, wifi_password) == ESP_OK) {
         g_wifi_connected = true;
         ESP_LOGI(TAG, "WiFi connected successfully");
         i2c_send_sound_command(SOUND_MELODY); // Send melody for success
@@ -117,17 +129,26 @@ void app_main(void) {
             ai_config_t ai_config;
 #if defined(CONFIG_AI_BACKEND_CLAUDE)
             ESP_LOGI(TAG, "Configuring Claude backend...");
-            ai_config.api_key = CLAUDE_API_KEY;
-            ai_config.api_url = CLAUDE_API_URL;
-            ai_config.model = CLAUDE_MODEL;
+            const char* claude_api_key = get_claude_api_key();
+            if (!claude_api_key || strlen(claude_api_key) == 0) {
+                ESP_LOGE(TAG, "Claude API key not available - cannot initialize Claude backend");
+                g_ai_backend = NULL;
+            } else {
+                ai_config.api_key = claude_api_key;
+                ai_config.api_url = "https://api.anthropic.com/v1/messages";
+                ai_config.model = "claude-3-haiku-20240307";
+            }
 #elif defined(CONFIG_AI_BACKEND_OLLAMA)
             ESP_LOGI(TAG, "Configuring Ollama backend...");
             ai_config.api_key = NULL; // Ollama doesn't require a key
             ai_config.api_url = OLLAMA_API_URL;
             ai_config.model = OLLAMA_MODEL;
 #endif
-            ESP_LOGI(TAG, "Calling backend initialization...");
-            esp_err_t init_result = g_ai_backend->init(&ai_config);
+            
+            // Only initialize if we have a valid backend
+            if (g_ai_backend) {
+                ESP_LOGI(TAG, "Calling backend initialization...");
+                esp_err_t init_result = g_ai_backend->init(&ai_config);
             if (init_result != ESP_OK) {
                 ESP_LOGE(TAG, "AI backend initialization FAILED with error: %s", esp_err_to_name(init_result));
 #if defined(CONFIG_AI_BACKEND_OLLAMA)
@@ -150,6 +171,7 @@ void app_main(void) {
             } else {
                 ESP_LOGI(TAG, "AI backend initialized successfully!");
                 ESP_LOGI(TAG, "System ready for AI-powered image analysis");
+                }
             }
         } else {
             ESP_LOGE(TAG, "CRITICAL: ai_backend_get_current() returned NULL!");

--- a/robocar/esp32-cam-idf/main/ollama_backend.c
+++ b/robocar/esp32-cam-idf/main/ollama_backend.c
@@ -34,12 +34,23 @@ static esp_err_t http_event_handler(esp_http_client_event_t *evt) {
             break;
         case HTTP_EVENT_ON_DATA:
             ESP_LOGD(TAG, "HTTP_EVENT_ON_DATA, len=%d", evt->data_len);
-            if (s_response_buffer && s_response_len + evt->data_len < MAX_RESPONSE_SIZE - 1) {
-                memcpy(s_response_buffer + s_response_len, evt->data, evt->data_len);
-                s_response_len += evt->data_len;
-                s_response_buffer[s_response_len] = '\0';
-            } else {
-                ESP_LOGW(TAG, "Response buffer overflow");
+            if (s_response_buffer && evt->data && evt->data_len > 0) {
+                // Ensure we don't overflow the buffer (MAX_RESPONSE_SIZE total, -1 for null terminator)
+                size_t available_space = MAX_RESPONSE_SIZE - 1 - s_response_len;
+                size_t bytes_to_copy = (evt->data_len <= available_space) ? evt->data_len : available_space;
+                
+                if (bytes_to_copy > 0) {
+                    memcpy(s_response_buffer + s_response_len, evt->data, bytes_to_copy);
+                    s_response_len += bytes_to_copy;
+                    s_response_buffer[s_response_len] = '\0';
+                    
+                    if (bytes_to_copy < evt->data_len) {
+                        ESP_LOGW(TAG, "HTTP response truncated: received %d bytes, only %zu bytes available", 
+                                evt->data_len, available_space);
+                    }
+                } else {
+                    ESP_LOGW(TAG, "HTTP response buffer full, discarding %d bytes", evt->data_len);
+                }
             }
             break;
         case HTTP_EVENT_ON_FINISH:

--- a/robocar/simulation/config/robot_config.yaml
+++ b/robocar/simulation/config/robot_config.yaml
@@ -28,7 +28,7 @@ robot:
       no_load_current: 0.1  # A
       stall_current: 2.0  # A
       resistance: 3.0  # Ohms
-      inductance: 0.001  # H
+      inductance: 0.01  # H (realistic for small DC motor)
       back_emf_constant: 0.01  # V⋅s/rad
       
     right:
@@ -41,7 +41,7 @@ robot:
       no_load_current: 0.1  # A
       stall_current: 2.0  # A
       resistance: 3.0  # Ohms
-      inductance: 0.001  # H
+      inductance: 0.01  # H (realistic for small DC motor)
       back_emf_constant: 0.01  # V⋅s/rad
 
   # Servo configuration (Camera mount)

--- a/robocar/simulation/requirements.txt
+++ b/robocar/simulation/requirements.txt
@@ -20,6 +20,7 @@ pyserial>=3.5
 
 # 3D graphics and visualization
 trimesh>=3.15.0
+swift-sim>=1.0.0
 # open3d>=0.15.0  # Not compatible with Python 3.13 yet
 # pyvista>=0.38.0  # Reduced for compatibility
 

--- a/robocar/simulation/src/robot_model.py
+++ b/robocar/simulation/src/robot_model.py
@@ -54,7 +54,8 @@ class DCMotor:
         self.resistance = config['resistance']  # Ohms
         self.inductance = config['inductance']  # H
         self.back_emf_constant = config['back_emf_constant']  # V⋅s/rad
-        self.torque_constant = config['back_emf_constant']  # N⋅m/A (same as back EMF)
+        # Torque constant derived from stall specifications for consistency
+        self.torque_constant = config['stall_torque'] / config['stall_current']  # N⋅m/A
         self.max_rpm = config['max_rpm']
         self.stall_torque = config['stall_torque']  # N⋅m
         self.no_load_current = config['no_load_current']  # A
@@ -65,10 +66,10 @@ class DCMotor:
         self.angular_velocity = 0.0  # rad/s
         self.torque = 0.0  # N⋅m
         
-        # Friction parameters
-        self.friction_static = 0.1  # N⋅m
-        self.friction_kinetic = 0.05  # N⋅m
-        self.friction_viscous = 0.001  # N⋅m⋅s/rad
+        # Friction parameters (realistic for small DC motor)
+        self.friction_static = 0.001  # N⋅m (reduced for small motor)
+        self.friction_kinetic = 0.0005  # N⋅m (reduced for small motor)
+        self.friction_viscous = 0.0001  # N⋅m⋅s/rad (reduced for small motor)
         
     def update(self, voltage: float, load_torque: float, dt: float) -> Tuple[float, float]:
         """
@@ -98,11 +99,14 @@ class DCMotor:
         motor_torque = self.torque_constant * self.current
         
         # Friction torque
-        if abs(self.angular_velocity) > 0.1:  # Moving
+        if abs(self.angular_velocity) > 0.01:  # Moving (reduced threshold)
             friction_torque = np.sign(self.angular_velocity) * self.friction_kinetic
-        else:  # Static or near-static
-            friction_torque = np.clip(motor_torque - load_torque, 
-                                    -self.friction_static, self.friction_static)
+        else:  # Static friction - only applies if motor torque exceeds static friction
+            static_threshold = abs(motor_torque - load_torque)
+            if static_threshold > self.friction_static:
+                friction_torque = np.sign(motor_torque - load_torque) * self.friction_static
+            else:
+                friction_torque = motor_torque - load_torque  # Static friction prevents motion
         
         # Viscous friction
         viscous_torque = self.friction_viscous * self.angular_velocity
@@ -110,12 +114,11 @@ class DCMotor:
         # Net torque
         net_torque = motor_torque - load_torque - friction_torque - viscous_torque
         
-        # Angular acceleration (assuming motor inertia is negligible compared to load)
-        # In practice, you'd include motor inertia: J * dw/dt = net_torque
-        if abs(net_torque) > 0.001:  # Avoid division by zero
-            # Simplified: assume small motor inertia
-            angular_acceleration = net_torque / 0.001  # kg⋅m²
-            self.angular_velocity += angular_acceleration * dt
+        # Angular acceleration with proper motor inertia
+        # J * dw/dt = net_torque, where J is motor rotor inertia
+        motor_inertia = 0.001  # kg⋅m² - realistic value for small DC motor rotor
+        angular_acceleration = net_torque / motor_inertia
+        self.angular_velocity += angular_acceleration * dt
         
         # Limit to maximum RPM
         max_angular_velocity = self.max_rpm * 2 * np.pi / 60  # rad/s

--- a/robocar/simulation/src/swift_visualizer.py
+++ b/robocar/simulation/src/swift_visualizer.py
@@ -9,6 +9,7 @@ import numpy as np
 import time
 import asyncio
 import threading
+import queue
 from typing import Dict, List, Optional, Tuple
 import yaml
 from pathlib import Path
@@ -34,7 +35,8 @@ class RobotVisualizer:
     
     def __init__(self, config_path: str, robot: DifferentialDriveRobot):
         if not HAS_SWIFT:
-            raise ImportError("Swift and Robotics Toolbox required for visualization")
+            print("Warning: Swift and Robotics Toolbox not available")
+            print("3D visualization will be disabled")
         
         self.config = self._load_config(config_path)
         self.robot = robot
@@ -48,6 +50,11 @@ class RobotVisualizer:
         self.wheel_meshes = []
         self.camera_mesh = None
         self.trail_points = []
+        
+        # Thread-safe communication
+        self._update_queue = queue.Queue()
+        self._running = False
+        self._thread = None
         self.obstacles = []
         
         # Visualization parameters
@@ -69,11 +76,10 @@ class RobotVisualizer:
             return yaml.safe_load(f)
     
     def _init_swift_env(self):
-        """Initialize Swift environment - disabled due to asyncio conflicts"""
-        # Temporarily disable Swift due to websockets/asyncio compatibility issues
-        print("Note: Swift 3D visualization temporarily disabled due to asyncio compatibility issues")
-        print("Running simulation in headless mode with text output")
-        self.env = None
+        """Initialize Swift environment with proper thread isolation"""
+        # We'll initialize Swift in the dedicated thread to avoid asyncio conflicts
+        self.env = None  # Will be created in _visualization_thread
+        print("Swift 3D visualization will be initialized in dedicated thread")
     
     def _create_environment(self):
         """Create the simulation environment"""
@@ -229,9 +235,25 @@ class RobotVisualizer:
         self.env.cam.set_pose(SE3(2, 2, 3) * SE3.Rz(-np.pi/4) * SE3.Rx(-np.pi/6))
     
     def update_robot_visualization(self):
-        """Update robot visualization based on current state"""
+        """Update robot visualization based on current state - thread safe"""
         state = self.robot.get_state()
         
+        # Send update command to visualization thread
+        update_data = {
+            'type': 'robot_update',
+            'state': state
+        }
+        try:
+            self._update_queue.put_nowait(update_data)
+        except queue.Full:
+            # Drop update if queue is full to prevent blocking
+            pass
+    
+    def _process_robot_update(self, state):
+        """Process robot visualization update in Swift thread"""
+        if not self.env:
+            return
+            
         # Update robot position and orientation
         robot_transform = SE3(state.x, state.y, 0.05) * SE3.Rz(state.theta)
         
@@ -274,6 +296,9 @@ class RobotVisualizer:
         
         # Update sensor visualization
         self._update_sensor_visualization(state)
+        
+        # Step the simulation
+        self.env.step(0.033)  # ~30 FPS
     
     def _update_trail(self, x: float, y: float):
         """Update robot trail visualization"""
@@ -345,12 +370,85 @@ class RobotVisualizer:
     
     def start_update_loop(self):
         """Start the visualization update loop"""
-        self.running = True
-        self.update_thread = threading.Thread(target=self._update_loop, daemon=True)
-        self.update_thread.start()
+        self._running = True
+        self._thread = threading.Thread(target=self._visualization_thread, daemon=True)
+        self._thread.start()
+    
+    def _visualization_thread(self):
+        """Main visualization thread - handles Swift environment isolated from asyncio"""
+        if not HAS_SWIFT:
+            print("Swift not available, visualization thread exiting")
+            return
+            
+        try:
+            # Create and set event loop for this thread
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            
+            print("Initializing Swift environment in dedicated thread...")
+            
+            # Suppress websocket error warnings that are internal to Swift
+            import warnings
+            import logging
+            
+            # Filter out websocket-related warnings
+            warnings.filterwarnings("ignore", message=".*websockets.*")
+            logging.getLogger("websockets").setLevel(logging.ERROR)
+            
+            # Initialize Swift environment with a unique port to avoid conflicts
+            import random
+            port = random.randint(8000, 9000)
+            
+            print(f"Note: Swift may show websocket threading warnings - these can be safely ignored")
+            self.env = Swift()
+            self.env.launch(realtime=True, headless=False, port=port)
+            print(f"✓ Swift 3D visualization running on port {port}")
+            print(f"✓ Open browser to http://localhost:{port} to view 3D visualization")
+            
+            # Create robot visualization elements
+            self._create_environment()
+            self._create_robot_model()
+            
+            # Run the visualization loop (websocket errors can be ignored)
+            last_robot_update = time.time()
+            update_interval = 1.0 / 30.0  # 30 FPS
+            
+            print("Starting Swift visualization loop...")
+            
+            while self._running:
+                current_time = time.time()
+                
+                # Process updates from queue
+                try:
+                    while True:
+                        update_data = self._update_queue.get_nowait()
+                        if update_data['type'] == 'robot_update':
+                            self._process_robot_update(update_data['state'])
+                        elif update_data['type'] == 'pose_update':
+                            self._process_robot_update(update_data['state'])
+                        # Add more update types as needed
+                except queue.Empty:
+                    pass
+                
+                # Regular robot state updates
+                if current_time - last_robot_update > update_interval:
+                    # Update with current robot state
+                    state = self.robot.get_state()
+                    self._process_robot_update(state)
+                    last_robot_update = current_time
+                
+                time.sleep(0.01)  # 100Hz check rate
+            
+            print("Swift visualization loop stopping")
+            
+        except Exception as e:
+            print(f"Swift initialization failed: {e}")
+            print("Running simulation without 3D visualization")
+            self.env = None
+            return
     
     def _update_loop(self):
-        """Main update loop for visualization"""
+        """Legacy update loop - now just calls update_robot_visualization"""
         while self.running:
             current_time = time.time()
             
@@ -362,11 +460,29 @@ class RobotVisualizer:
     
     def stop(self):
         """Stop the visualization"""
-        self.running = False
-        if self.update_thread:
-            self.update_thread.join()
+        print("Stopping Swift visualization...")
+        self._running = False
+        self.running = False  # Legacy compatibility
+        
+        # Stop the event loop in the visualization thread
+        if hasattr(self, '_thread') and self._thread and self._thread.is_alive():
+            # Send a signal to stop the event loop
+            try:
+                loop = asyncio.get_running_loop()
+                if loop:
+                    loop.call_soon_threadsafe(loop.stop)
+            except:
+                pass
+            self._thread.join(timeout=2.0)  # Wait up to 2 seconds
+        
+        if hasattr(self, 'update_thread') and self.update_thread:
+            self.update_thread.join(timeout=1.0)
+            
         if self.env:
-            self.env.close()
+            try:
+                self.env.close()
+            except Exception as e:
+                print(f"Warning: Error closing Swift environment: {e}")
     
     def screenshot(self, filename: str):
         """Take a screenshot of the visualization"""

